folio-token/src/test/java/org/folio/spring/systemuser/SystemUserServiceTest.java

```diff
@@ -0,0 +1,213 @@
+
+@ExtendWith(MockitoExtension.class)
+class SystemUserServiceTest {
+
+  public static final String OKAPI_URL = "http://okapi";
+  private static final String TENANT_ID = "test";
+  private static final Instant TOKEN_EXPIRATION = Instant.now().plus(1, ChronoUnit.DAYS);
+  private static final String MOCK_TOKEN = "test_token";
+  @Mock
+  private AuthnClient authnClient;
+  @Mock
+  private ExecutionContextBuilder contextBuilder;
+  private final ResponseEntity<AuthnClient.LoginResponse> expectedResponse =
+      Mockito.spy(ResponseEntity.of(Optional.of(new AuthnClient.LoginResponse(TOKEN_EXPIRATION.toString()))));
+  @Mock
+  private FolioExecutionContext context;
+  @Mock
+  private FolioEnvironment environment;
+  @Mock
+  private PrepareSystemUserService prepareSystemUserService;
+  @Mock
+  private Cache<String, SystemUser> userCache;
+
+  private static SystemUser systemUserValue() {
+    return SystemUser.builder().username("username").okapiUrl(OKAPI_URL).tenantId(TENANT_ID).build();
+  }
+
+  private static SystemUserProperties systemUserProperties() {
+    return new SystemUserProperties("username", "password", "system", "permissions/test-permissions.csv");
+  }
+
+  @Test
+  void getAuthedSystemUser_positive() {
+    var expectedUserId = UUID.randomUUID();
+    var expectedUserToken = userToken(TOKEN_EXPIRATION);
+    var expectedHeaders = cookieHeaders(expectedUserToken.accessToken(), expectedUserToken.accessToken());
+
+    when(authnClient.loginWithExpiry(new UserCredentials("username", "password"))).thenReturn(expectedResponse);
+    when(prepareSystemUserService.getFolioUser("username")).thenReturn(Optional.of(
+        new UsersClient.User(expectedUserId.toString(), "username", true, new UsersClient.User.Personal("last"))));
+    when(environment.getOkapiUrl()).thenReturn(OKAPI_URL);
+    when(contextBuilder.forSystemUser(any())).thenReturn(context);
+    when(expectedResponse.getHeaders()).thenReturn(expectedHeaders);
+
+    var actual = systemUserService(systemUserProperties()).getAuthedSystemUser(TENANT_ID);
+    assertThat(actual.token()).isEqualTo(expectedUserToken);
+    assertThat(actual.userId()).isEqualTo(expectedUserId.toString());
+  }
+
+  @Test
+  void getAuthedSystemUserUsingCache_positive() {
+    var expectedUserToken = userToken(Instant.now().plus(1, ChronoUnit.DAYS));
+    var systemUserService = systemUserService(systemUserProperties());
+    systemUserService.setSystemUserCache(userCache);
+
+    when(userCache.get(eq(TENANT_ID), any())).thenReturn(systemUserValue().withToken(expectedUserToken));
+
+    var actual = systemUserService.getAuthedSystemUser(TENANT_ID);
+    assertThat(actual.token().accessToken()).isEqualTo(expectedUserToken.accessToken());
+    verify(userCache).get(eq(TENANT_ID), any());
+    verify(authnClient, never()).loginWithExpiry(any());
+    verify(environment, never()).getOkapiUrl();
+    verify(contextBuilder, never()).forSystemUser(any());
+  }
+
```

Comment:
@nirmalr-shah I think we need a `getAuthedSystemUserUsingCache_negative` test that adds an _expired_ token to the cache, and then verifies that the logic we expect in `SystemUserService.getAuthedSystemUser` gets called. When the token is expired it will be before `Instant.now` and `systemUserCache.invalidate` should be called and login should occur.

Reply:
added the test case, please review!!